### PR TITLE
Review tests

### PR DESCRIPTION
**Pull request by emilejq**
## Proposed changes

This PR addresses the open issue to review the tests

**General Changes**
- The tests have been separated into unit and integration tests, and the directory structure has been updated accordingly
- The `test_rules.yml` file has been deleted and all the tests just instead use the `default_rules.yml` file in the test files
- An unused `bulk_headers.json` file has been removed from the test files
- Any places where a resource opened within a `with` statement is being explicitly closed have been removed. The clean up of such resources is implicit in the `with` statement

**Integration Tests**
- The integration tests are separated into two classes:
- `TestDrHeader` tests the logic of the main code. Each test in this class first modifies the headers and rules used for the test so that the pre-requisites to successfully run the test are in place. This is done so that these tests are resilient to any changes to the default rules, such as changing the accepted value for a header, or a header being deprecated.
- `TestDefaultRules` contains the tests that verify the rules and expected values of the headers. Changes to the rules or headers will be picked by the tests in this class without affecting other integration tests
- Both integration test classes inherit from a base class that contains shared methods
- All the integration tests (except one) now use the `build_error_message` method for better error reporting. Additionally, this method has been modified to only report discrepancies in the report about the specific rule being tested - unrelated unexpected items in the report will be ignored when creating the error message

**Unit Tests**
- Unit tests for the CLI have been added
- The existing unit tests `TestCliUtils` have been moved into a shared module with the new unit tests for the CLI. These unit tests are unmodified otherwise
- Some changes to the unit tests for the utils class to tidy up the tests/coverage and add more descriptive names to the tests

**Changes to the main code**
- An unused `_to_lower_dict` function has been removed from the utils class
- The `X-Powered-By` header has been capitalised to be consistent with the others
- A spelling error and an uppercase parameter that were causing complaints in the IDE have been changed

## Type of change

What types of changes do you want to introduce to DrHeader?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change 
- [ ] Documentation Update
- [x] Test Update
- [ ] Rules Update 
- [ ] Other (please describe)

Please ensure your pull request adheres to the following guidelines:
- [ ] A Github Issue that explains the work.
- [ ] The changes are in a branch that is reasonably up to date
- [ ] Tests are provided to reasonably cover new or altered functionality.
- [ ] Documentation is provided for the new or altered functionality.
- [ ] You have the legal right to give us this code.
- [ ] You have adhered to the CoC

## Link to the github issue 

https://github.com/Santandersecurityresearch/DrHeader/issues/172

## Tests you have added 

Many

## Tests you have altered

Many

## Anything Else 

Thanks for getting this far !

### CODE REVIEW

**Review by mrjonstrong — COMMENTED** [review]
Only minor points, you did ask to take a look at the pythonness of it. I think technically some long lines in there too as PEP 8 Style guide says to limit to 79 characters.

File: tests/integration_tests/test_base.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import unittest2
+import yaml
+
+from drheader import Drheader
+
+
+class TestBase(unittest2.TestCase):
+
+    def setUp(self):
```
[the remark this answers is not in the document]
Comment (emilejq) [reply tests/integration_tests/test_base.py:14]: This method overrides a method of the same name in the unittest2.TestCase class, hence the method signature is required to match

File: tests/integration_tests/test_base.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        self.logger = logging.Logger
+
+    def tearDown(self):
```
[the remark this answers is not in the document]
Comment (emilejq) [reply tests/integration_tests/test_base.py:17]: This method overrides a method of the same name in the unittest2.TestCase class, hence the method signature is required to match

[the remark this answers is not in the document]
Comment (emilejq) [reply tests/integration_tests/test_base.py:17]: Nope, as the method signature should match that of the base method again

**Review by mrjonstrong — APPROVED** [review]

**Pull request merged**